File: astar.ts
// Typescriptified version of https://github.com/superjoe30/node-astar

import Heap = require('heap')

type Node<T> = {
  data: T
  g: number
  h: number
  f: number
  parent: Node<T>
}

type Params<T> = {
  start: T
  isEnd: (node: T) => boolean
  neighbors: (node: T) => T[]
  cost?: (a: T, b: T) => number
  heuristic?: (node: T) => number
  hash?: (node: T) => string
}

type Result<T> = {
  cost: number
  path: T[]
}

export default function aStar<T>(params: Params<T>): Result<T> {
  const hash = params.hash || JSON.stringify
  const cost = params.cost || (() => 1)
  const heuristic = params.heuristic || (() => 1)

  const startNode: Node<T> = {
    data: params.start,
    g: 0,
    h: heuristic(params.start),
    f: 0,
    parent: null
  }
  startNode.f = startNode.h

  const closedDataSet = new Set<string>()
  const openHeap = new Heap<Node<T>>((a, b) => a.f - b.f)
  const openDataMap = new Map<string, Partial<Node<T>>>()

  openHeap.push(startNode)
  openDataMap.set(hash(startNode.data), startNode)
  while (!openHeap.empty()) {
    const node = openHeap.pop()
    const nodeHash = hash(node.data)
    openDataMap.delete(nodeHash)

    if (params.isEnd(node.data)) {
      // done
      return {
        cost: node.g,
        path: reconstructPath(node)
      }
    }

    // not done yet
    closedDataSet.add(nodeHash)

    for (const neighborData of params.neighbors(node.data)) {
      const neighborHash = hash(neighborData)
      if (closedDataSet.has(neighborHash)) {
        // skip closed neighbors
        continue
      }
      const gFromThisNode = node.g + cost(node.data, neighborData)
      let neighborNode = openDataMap.get(neighborHash)
      let update = false

      if (!neighborNode) {
        neighborNode = {
          data: neighborData
          // other properties will be set later
        }
        // add neighbor to the open set
        openDataMap.set(neighborHash, neighborNode)
      } else {
        if (neighborNode.g < gFromThisNode) {
          // skip this one because another route is faster
          continue
        }
        update = true
      }

      // found a new or better route.
      // update this neighbor with this node as its new parent
      neighborNode.parent = node
      neighborNode.g = gFromThisNode
      neighborNode.h = heuristic(neighborData)
      neighborNode.f = gFromThisNode + neighborNode.h

      if (update) {
        openHeap.heapify()
      } else {
        openHeap.push(neighborNode as Node<T>)
      }
    }
  }
  // all the neighbors of every accessible node have been exhausted
  return null
}

const reconstructPath = <T>(node: Node<T>): T[] =>
  node.parent ? [...reconstructPath(node.parent), node.data] : [node.data]

File: astar.clj
(ns aoc.astar
  (:require [clojure.data.priority-map :refer [priority-map]]))

(defn- reconstruct-path [node]
  (cons (:data node)
        (when (some? (:parent node))
          (lazy-seq (reconstruct-path (:parent node))))))

(defn astar [& {:keys [start is-end get-neighbors calculate-cost heuristic hash validate-path]
                :or {calculate-cost (constantly 1)
                     heuristic (constantly 1)
                     hash hash}}]

  (loop [closed-map {}
         open-map {}
         priorities (priority-map)
         node {:data start
               :g 0
               :h (heuristic start)
               :parent nil}]

    (cond
      ;; Did not find a path
      (or (nil? node)
          (and (= 0 (count priorities))
               (> 0 (count closed-map)))) nil
      ;; Done!
      (is-end (:data node)) {:cost (:g node) :path (reconstruct-path node) :visited (->> closed-map (map (fn [[_ v]] [(:data v) v])) (into {}))}
      ;; Not done yet
      :else
      (let [hashed-data (hash (:data node))
            closed-map (assoc closed-map hashed-data (dissoc node :parent))
            open-map (dissoc open-map hashed-data)
            [priorities open-map]
            (->> (get-neighbors (:data node))
                 (filter #(not (contains? closed-map (hash %))))
                 ;; Validate path to this neighbor if we have a validate-path-function to call
                 (filter #(if (some? validate-path)
                            (validate-path (reconstruct-path {:data % :parent node}))
                            true))
                 (reduce
                  (fn [[priorities open-map] neighbor-data]
                    (let [neighbor-hash (hash neighbor-data)
                          g-from-this-node (+ (:g node) (calculate-cost (:data node) neighbor-data))]
                      (if
                       ;; We have seen this node before and already have a faster path to it
                       (and (contains? open-map neighbor-hash)
                            (< (get-in open-map [neighbor-hash :g]) g-from-this-node))
                        [priorities open-map]

                        (let [new-neighbor-node {:data neighbor-data
                                                 :parent node
                                                 :g g-from-this-node
                                                 :h (heuristic neighbor-data)}]
                          (if-let [old-neighbor-node (get open-map neighbor-hash)]
                            ;; Found a better route, update neighbor with this node as its new parent
                            [(-> priorities
                                 (dissoc old-neighbor-node)
                                 (assoc new-neighbor-node (+ g-from-this-node (:h new-neighbor-node))))
                             (assoc open-map neighbor-hash new-neighbor-node)]
                            ;; Add new neighbor to open nodes
                            [(assoc priorities new-neighbor-node (+ g-from-this-node (:h new-neighbor-node)))
                             (assoc open-map neighbor-hash new-neighbor-node)])))))
                  [priorities open-map]))]

        (recur closed-map
               open-map
               (if (= 0 (count priorities)) priorities (pop priorities))
               (first (peek priorities)))))))
